code-standards/php.md

```diff
@@ -71,6 +72,155 @@ rules:
+
+    #### Interface Layer
+
+    No functionality whatsoever may go into this layer. Interface packages
+    may depend on other interface packages. Depending on packages containing
+    functionality is strongly discouraged. Tests must verify that classes
+    implementing the interfaces can be created, and that they implement all
+    other required interfaces - which may be a consequence of an interface
+    hierarchy. Such packages names MUST be suffixed with `-interface` -
+    regardless of the amount of interfaces declared.
+
+    #### Abstract Layer
+
+    No concrete implementations may go into this layer. Only abstract classes
+    are allowed. All methods must not be declared as `public`. Static
+    methods are advised against to preserve IoC principle. Methods which
+    are meant to create instances must be `abstract`. A very high standard
+    of granularity should be maintained. Only generic functionality can be
+    implemented at this layer. Abstract classes MUST NOT implement interfaces,
+    or declare a public constructor. If descendants are anticipated to
+    implement specific interfaces, its methods or constants MUST NOT be
+    referred to as those belonging to the abstract class, i.e. by using the
+    `static` keyword, or in any other way. Instead, only the protected
+    methods of the abstract class itself may be invoked. Interface constants
+    MUST be referred to by using their fully qualified name, i.e.
+    `Interface::CONSTANT`. It is possible to use the `use` keyword to alias
+    other classes and interfaces to shorter symbols. No other functionality,
+    except for that in the package itself and its dependencies, can be
+    assumed. The exception to this rule is typechecking, such as in cases
+    where additional processing should be performed for specific types,
+    or when certain exceptions need to be handled. This is because checking
+    for types, explicitly or for exception handling, do not cause fatal
+    errors if the checked type is unavailable, i.e. the existence of the
+    type itself is not required. This layer MUST remain completely agnostic
+    of any particular use-case, environment, or application. Such package names
+    MUST be suffixed with `-abstract`.
+
+    #### Base Layer (optional)
+
+    This level allows creation of concrete functionality, declaration of
+    public methods, and concreate classes. However, this is done only
+    for the purpose of saving time when coding concrete implementations
+    that frequently use the most common functionality, such as interface
+    (public) methods, or exceptions. Classes SHOULD be declared as
+    `abstract` even if they have enough data or functionality to be
+    instantiated. Such classes MUST be prefixed with the word "Base",
+    in addition to the PSR naming conventions mentioned earlier in this
+    document. This means that for a base class which excents
```

Comment:
Possible typo: "excents" instead of "extends". Unless "excents" is a word I've never heard of.